File: karaoke-mobile/lib/player/playback_engine.dart
import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:just_audio/just_audio.dart';
import 'package:video_player/video_player.dart';

import '../api/client.dart';
import '../api/media_cache.dart';
import '../api/models.dart';

enum Track { original, instrumental }

/// Plays the original and instrumental tracks in parallel and crossfades their
/// gains so switching to "karaoke" is seamless (no seek, no gap). The audible
/// track is the clock; the muted track and the (muted) MV are nudged back into
/// sync on drift. Mirrors the desktop PlaybackEngine.
class PlaybackEngine extends ChangeNotifier {
  final _original = AudioPlayer();
  final _instrumental = AudioPlayer();
  VideoPlayerController? _video;

  Track _track = Track.original;
  bool _hasOriginal = false;
  bool _hasInstrumental = false;
  bool _videoLoops = false; // MV much shorter than audio -> repeats
  Duration _duration = Duration.zero;
  Timer? _driftTimer;
  Timer? _fadeTimer;

  Track get track => _track;
  bool get canToggle => _hasOriginal && _hasInstrumental;
  bool get hasMedia => _hasOriginal || _hasInstrumental;
  VideoPlayerController? get video => _video;
  Duration get duration => _duration;

  // The audible track is the master clock (instrumental is always present when
  // toggling; otherwise whichever single track loaded).
  AudioPlayer get _master => _hasInstrumental ? _instrumental : _original;

  // Single source of truth for play state, so the button icon and togglePlay()
  // never disagree (the underlying players' `playing` can lag a stream tick,
  // which made the first resume tap a no-op).
  bool _playing = false;
  bool get playing => _playing;

  /// Wall-clock-interpolated position (seconds) — smooth enough for the wipe.
  double get positionSec => _master.position.inMicroseconds / 1e6;

  Future<void> load(ApiClient api, Song song) async {
    await _reset();
    _hasOriginal = song.hasOriginal;
    _hasInstrumental = song.hasInstrumental;
    // Default to vocals; the user toggles to karaoke. If only an instrumental
    // exists, play that.
    _track = (_hasOriginal) ? Track.original : Track.instrumental;

    // Serve from the on-device cache when a fresh copy exists (validated
    // against the server's ETag, so replaced files auto-invalidate); otherwise
    // stream from the network while the cache fills in the background.
    final cache = await MediaCache.open();
    Future<Duration?> loadAudio(AudioPlayer player, String track) async {
      final url = api.audioUrl(song.id, track);
      // Audio always plays from a local file (downloaded now if needed):
      // seeking a network stream can stall ExoPlayer indefinitely, and audio
      // is small enough that a first-play LAN download takes ~1-3s.
      final file = await cache.fetch('${song.id}.$track', url);
      return file != null ? player.setFilePath(file.path) : player.setUrl(url);
    }

    final durs = await Future.wait([
      if (_hasOriginal) loadAudio(_original, 'original'),
      if (_hasInstrumental) loadAudio(_instrumental, 'instrumental'),
    ]);
    for (final d in durs) {
      if (d != null && d > _duration) _duration = d;
    }

    if (song.hasVideo) {
      // Best-effort: a video that can't load (server down, not yet cached)
      // must degrade to audio-only playback, never block the song.
      try {
        final url = api.videoUrl(song.id);
        final file = await cache.resolve('${song.id}.video.mp4', url);
        // mixWithOthers: the muted MV must never touch Android audio focus —
        // focus grabs/abandons at its play/complete transitions pause the
        // just_audio tracks (heard as playback freezing at the video's end).
        final opts = VideoPlayerOptions(mixWithOthers: true);
        final v = file != null
            ? VideoPlayerController.file(file, videoPlayerOptions: opts)
            : VideoPlayerController.networkUrl(Uri.parse(url), videoPlayerOptions: opts);
        await v.initialize();
        await v.setVolume(0); // audio comes from the audio players
        // MV shorter than the audio: loop it when the audio is much longer
        // (>1.5x); otherwise it ends and freezes on the last frame.
        final vd = v.value.duration;
        _videoLoops = vd > Duration.zero &&
            _duration.inMilliseconds > 1.5 * vd.inMilliseconds;
        await v.setLooping(_videoLoops);
        _video = v;
      } catch (_) {
        _video = null;
      }
    }

    _applyVolumes(immediate: true);
    notifyListeners();
  }

  void play() {
    _playing = true;
    if (_hasOriginal) _original.play(); // NB: just_audio play() completes at EOF
    if (_hasInstrumental) _instrumental.play();
    _video?.play();
    _startDrift();
    notifyListeners();
  }

  Future<void> pause() async {
    _playing = false;
    _stopDrift();
    await Future.wait([
      if (_hasOriginal) _original.pause(),
      if (_hasInstrumental) _instrumental.pause(),
      if (_video != null) _video!.pause(),
    ]);
    notifyListeners();
  }

  void togglePlay() => playing ? pause() : play();

  /// Where the MV should be for audio position [pos] (modulo when looping,
  /// clamped to the last frame otherwise).
  Duration _videoTarget(Duration pos) {
    final vd = _video?.value.duration ?? Duration.zero;
    if (vd <= Duration.zero) return pos;
    if (_videoLoops) {
      return Duration(milliseconds: pos.inMilliseconds % vd.inMilliseconds);
    }
    return pos < vd ? pos : vd;
  }

  Future<void> seek(Duration pos) async {
    await Future.wait([
      if (_hasOriginal) _original.seek(pos),
      if (_hasInstrumental) _instrumental.seek(pos),
    ]);
    final v = _video;
    if (v != null && v.value.isInitialized) {
      final vd = v.value.duration;
      final beyondEnd = !_videoLoops && vd > Duration.zero && pos >= vd;
      if (beyondEnd) {
        // Hold the final frame; never "play" a completed video (its restart/
        // complete transitions are what cause trouble).
        await v.seekTo(vd - const Duration(milliseconds: 100));
        await v.pause();
      } else {
        await v.seekTo(_videoTarget(pos));
        if (_playing) await v.play(); // was frozen at the end? moving again
      }
    }
  }

  Future<void> seekBy(int seconds) =>
      seek(_master.position + Duration(seconds: seconds));

  Future<void> restart() async {
    await seek(Duration.zero);
    play();
  }

  void setTrack(Track t) {
    if (!canToggle || t == _track) return;
    _track = t;
    _crossfade();
    notifyListeners();
  }

  void toggleTrack() =>
      setTrack(_track == Track.original ? Track.instrumental : Track.original);

  // ---- internals ---------------------------------------------------------

  void _applyVolumes({bool immediate = false}) {
    final origTarget = _track == Track.original ? 1.0 : 0.0;
    final instTarget = _track == Track.instrumental ? 1.0 : 0.0;
    if (immediate) {
      if (_hasOriginal) _original.setVolume(_hasInstrumental ? origTarget : 1.0);
      if (_hasInstrumental) _instrumental.setVolume(_hasOriginal ? instTarget : 1.0);
    }
  }

  void _crossfade() {
    _fadeTimer?.cancel();
    const steps = 8;
    const total = Duration(milliseconds: 120);
    final origStart = _original.volume, instStart = _instrumental.volume;
    final origTarget = _track == Track.original ? 1.0 : 0.0;
    final instTarget = _track == Track.instrumental ? 1.0 : 0.0;
    var i = 0;
    _fadeTimer = Timer.periodic(total ~/ steps, (tmr) {
      i++;
      final k = (i / steps).clamp(0.0, 1.0);
      _original.setVolume(origStart + (origTarget - origStart) * k);
      _instrumental.setVolume(instStart + (instTarget - instStart) * k);
      if (i >= steps) tmr.cancel();
    });
  }

  void _startDrift() {
    _stopDrift();
    // Gentle, infrequent correction. Hard-seeking the (network-streamed) video
    // on small drift causes rebuffer stutter, so only correct large drift.
    _driftTimer = Timer.periodic(const Duration(milliseconds: 1000), (_) {
      if (!_playing) return;
      final m = _master.position;
      if (canToggle) {
        final other = _hasInstrumental ? _original : _instrumental;
        if ((other.position - m).inMilliseconds.abs() > 400) other.seek(m);
      }
      final v = _video;
      if (v != null && v.value.isInitialized) {
        final vd = v.value.duration;
        if (!_videoLoops && vd > Duration.zero && m >= vd) {
          // MV over, audio continues: frozen last frame by design.
        } else {
          final target = _videoTarget(m);
          var deltaMs = (v.value.position - target).inMilliseconds;
          if (_videoLoops && vd > Duration.zero) {
            // Wrap-around distance: don't hard-seek across the loop seam.
            final half = vd.inMilliseconds ~/ 2;
            if (deltaMs > half) deltaMs -= vd.inMilliseconds;
            if (deltaMs < -half) deltaMs += vd.inMilliseconds;
          }
          if (deltaMs.abs() > 800) v.seekTo(target);
        }
      }
    });
  }

  void _stopDrift() {
    _driftTimer?.cancel();
    _driftTimer = null;
  }

  Future<void> _reset() async {
    _playing = false;
    _videoLoops = false;
    _stopDrift();
    _fadeTimer?.cancel();
    await _original.stop();
    await _instrumental.stop();
    final v = _video;
    _video = null;
    await v?.dispose();
    _hasOriginal = _hasInstrumental = false;
    _duration = Duration.zero;
  }

  /// Fully stop and release the current song (engine stays reusable).
  Future<void> stop() async {
    await _reset();
    notifyListeners();
  }

  @override
  void dispose() {
    _stopDrift();
    _fadeTimer?.cancel();
    _original.dispose();
    _instrumental.dispose();
    _video?.dispose();
    super.dispose();
  }
}

File: karaoke-ios/karaoke/Player/PlaybackEngine.swift
import AVFoundation
import Foundation
import Observation

enum AudioTrack: String {
    case original, instrumental
}

/// Plays the original and instrumental tracks in parallel and crossfades their
/// gains, so switching to "karaoke" is seamless — no seek, no gap. The audible
/// track is the master clock; the silent track and the (muted) MV are pulled
/// back into sync when they drift. Mirrors the desktop and Flutter engines.
@Observable
@MainActor
final class PlaybackEngine {
    private let original = AVPlayer()
    private let instrumental = AVPlayer()
    /// Handed to the video layer view; nil when the song has no MV.
    private(set) var video: AVPlayer?

    private(set) var track: AudioTrack = .original
    private(set) var isPlaying = false
    private(set) var duration: Double = 0
    private(set) var isLoading = false
    private(set) var loadError: String?
    /// Set when a song loads in a degraded way (one track undecodable), so the
    /// player can say why the karaoke toggle is missing.
    private(set) var trackNotice: String?

    private var hasOriginal = false
    private var hasInstrumental = false
    /// MV much shorter than the audio: repeat it instead of freezing.
    private var videoLoops = false
    private var videoDuration: Double = 0
    private var videoOffsetSec: Double = 0

    private var driftTimer: Timer?
    /// Drift is ignored until this instant, so a correction is not immediately
    /// re-judged while the decoder is still catching up.
    private var videoSettledAt: Date = .distantPast
    private var fadeTimer: Timer?
    private var observers: [NSObjectProtocol] = []
    private var loadToken = 0

    /// An explicit toggle is a standing choice that carries to the next song.
    var trackPreference: AudioTrack = .original

    /// Fired when the master track reaches its end (queue auto-advance).
    var onEnded: (() -> Void)?

    var canToggle: Bool { hasOriginal && hasInstrumental }
    var hasMedia: Bool { hasOriginal || hasInstrumental }

    /// The audible track is the clock: the instrumental when both exist,
    /// otherwise whichever single track loaded.
    private var master: AVPlayer { hasInstrumental ? instrumental : original }
    private var audioPlayers: [AVPlayer] {
        (hasOriginal ? [original] : []) + (hasInstrumental ? [instrumental] : [])
    }

    var positionSec: Double {
        let t = master.currentTime().seconds
        return t.isFinite ? max(0, t) : 0
    }

    init() {
        for p in [original, instrumental] {
            // Required before setRate(_:time:atHostTime:), which is how both
            // tracks are started on the same host clock tick.
            p.automaticallyWaitsToMinimizeStalling = false
            p.volume = 0
        }
    }

    // ---- loading -----------------------------------------------------------

    func load(api: APIClient, song: Song) async {
        await reset()
        loadToken += 1
        let token = loadToken
        isLoading = true
        loadError = nil
        trackNotice = nil

        hasOriginal = song.hasOriginal
        hasInstrumental = song.hasInstrumental
        guard hasMedia else {
            isLoading = false
            loadError = "this song has no audio on the server"
            return
        }
        // Default to the standing original/karaoke choice; fall back to
        // whichever track actually exists.
        track = trackPreference
        if track == .original && !hasOriginal { track = .instrumental }
        if track == .instrumental && !hasInstrumental { track = .original }
        videoOffsetSec = Double(song.videoOffsetMs) / 1000

        if hasOriginal {
            original.replaceCurrentItem(with: AVPlayerItem(url: api.audioURL(song.id, track: "original")))
        }
        if hasInstrumental {
            instrumental.replaceCurrentItem(with: AVPlayerItem(url: api.audioURL(song.id, track: "instrumental")))
        }
        applyGains(immediate: true)

        if song.hasVideo {
            let p = AVPlayer(url: api.videoURL(song.id))
            p.automaticallyWaitsToMinimizeStalling = false
            p.isMuted = true // all audio comes from the audio players
            video = p
        }

        // Wait until every player can actually start, so the transport is not
        // live before the first frame/sample is there. A track the device
        // cannot decode (Opus, say) is dropped here rather than left in place
        // as a dead master clock that freezes the whole song at 0:00.
        if hasOriginal, await !Self.waitUntilReady(original) {
            hasOriginal = false
            original.replaceCurrentItem(with: nil)
        }
        if hasInstrumental, await !Self.waitUntilReady(instrumental) {
            hasInstrumental = false
            instrumental.replaceCurrentItem(with: nil)
        }
        guard hasMedia else {
            isLoading = false
            loadError = "This iPad can't play this song's audio format."
            return
        }
        if track == .instrumental && !hasInstrumental { track = .original }
        if track == .original && !hasOriginal { track = .instrumental }
        if !canToggle {
            trackNotice = hasInstrumental
                ? "Only the karaoke track is playable on iPad."
                : "Only the original track is playable on iPad — no karaoke mode for this song."
        }
        applyGains(immediate: true)
        if let v = video {
            // A video that won't load must degrade to audio-only playback,
            // never block the song.
            if await Self.waitUntilReady(v), let item = v.currentItem {
                let d = item.duration.seconds
                videoDuration = d.isFinite ? d : 0
            } else {
                video = nil
            }
        }
        guard token == loadToken else { return } // superseded by a newer load

        duration = audioPlayers
            .compactMap { $0.currentItem?.duration.seconds }
            .filter { $0.isFinite && $0 > 0 }
            .max() ?? (song.durationSec ?? 0)
        videoLoops = videoDuration > 0 && duration > 1.5 * videoDuration
        installEndObservers()
        isLoading = false
        if !hasMedia { loadError = "nothing to play" }
    }

    /// Polls `status` — simpler than KVO and the wait is bounded.
    private static func waitUntilReady(_ player: AVPlayer, timeout: Double = 20) async -> Bool {
        let deadline = Date().addingTimeInterval(timeout)
        while Date() < deadline {
            switch player.currentItem?.status {
            case .readyToPlay: return true
            case .failed: return false
            default: try? await Task.sleep(for: .milliseconds(50))
            }
        }
        return false
    }

    // ---- transport ---------------------------------------------------------

    func play() {
        guard hasMedia, !isLoading else { return }
        isPlaying = true
        // Start every player on the same host-clock tick so the tracks (and the
        // MV) line up from the first sample instead of converging afterwards.
        videoRate = 1
        videoSettledAt = Date().addingTimeInterval(1.5)
        let host = CMClockGetTime(CMClockGetHostTimeClock())
            + CMTime(seconds: 0.15, preferredTimescale: 600)
        for p in audioPlayers { p.setRate(1, time: .invalid, atHostTime: host) }
        if let v = video, !(videoEnded(at: positionSec)) {
            v.setRate(1, time: .invalid, atHostTime: host)
        }
        startDrift()
    }

    func pause() {
        isPlaying = false
        videoRate = 1
        stopDrift()
        for p in audioPlayers { p.pause() }
        video?.pause()
    }

    func togglePlay() { isPlaying ? pause() : play() }

    func seek(to seconds: Double) {
        let clamped = max(0, duration > 0 ? min(seconds, duration) : seconds)
        let time = CMTime(seconds: clamped, preferredTimescale: 600)
        for p in audioPlayers {
            p.seek(to: time, toleranceBefore: .zero, toleranceAfter: .zero)
        }
        syncVideo(to: clamped, tolerance: 0.05, resume: true)
    }

    func seekBy(_ delta: Double) { seek(to: positionSec + delta) }

    func restart() {
        seek(to: 0)
        play()
    }

    func setTrack(_ t: AudioTrack) {
        guard canToggle, t != track else { return }
        track = t
        trackPreference = t
        applyGains(immediate: false)
    }

    func toggleTrack() { setTrack(track == .original ? .instrumental : .original) }

    /// Fully stop and release the current song; the engine stays reusable.
    func stop() async { await reset() }

    // ---- gains -------------------------------------------------------------

    private func applyGains(immediate: Bool) {
        // With only one track loaded it is always audible, whatever `track` says.
        let originalTarget: Float = canToggle ? (track == .original ? 1 : 0) : 1
        let instrumentalTarget: Float = canToggle ? (track == .instrumental ? 1 : 0) : 1
        fadeTimer?.invalidate()
        guard !immediate else {
            original.volume = hasOriginal ? originalTarget : 0
            instrumental.volume = hasInstrumental ? instrumentalTarget : 0
            return
        }
        let steps = 8
        let originalStart = original.volume, instrumentalStart = instrumental.volume
        var i = 0
        fadeTimer = Timer.scheduledTimer(withTimeInterval: 0.12 / Double(steps), repeats: true) { [weak self] timer in
            MainActor.assumeIsolated {
                guard let self else { return timer.invalidate() }
                i += 1
                let k = Float(min(1, Double(i) / Double(steps)))
                self.original.volume = originalStart + (originalTarget - originalStart) * k
                self.instrumental.volume = instrumentalStart + (instrumentalTarget - instrumentalStart) * k
                if i >= steps { timer.invalidate() }
            }
        }
    }

    // ---- A/V sync ----------------------------------------------------------

    /// Where the MV should be for audio time `t`. A positive video offset runs
    /// the picture ahead of the audio, so lyrics burned into it arrive earlier.
    private func videoTarget(for t: Double) -> Double {
        guard videoDuration > 0 else { return t }
        let shifted = t + videoOffsetSec
        if videoLoops { return shifted.truncatingRemainder(dividingBy: videoDuration) + (shifted < 0 ? videoDuration : 0) }
        return min(max(shifted, 0), videoDuration)
    }

    /// True once a non-looping MV is over but the audio keeps going — the last
    /// frame stays on screen by design.
    private func videoEnded(at t: Double) -> Bool {
        !videoLoops && videoDuration > 0 && t + videoOffsetSec >= videoDuration
    }

    /// Video seeks are always tolerant: an exact seek has to decode every frame
    /// from the preceding keyframe, which on a busy decoder costs more time
    /// than the drift it was correcting — the stutter feeds itself.
    private func syncVideo(to t: Double, tolerance: Double, resume: Bool) {
        guard let v = video else { return }
        if videoEnded(at: t) {
            v.seek(to: CMTime(seconds: max(0, videoDuration - 0.1), preferredTimescale: 600))
            v.pause()
            return
        }
        let target = CMTime(seconds: videoTarget(for: t), preferredTimescale: 600)
        let slack = CMTime(seconds: tolerance, preferredTimescale: 600)
        v.seek(to: target, toleranceBefore: slack, toleranceAfter: slack)
        videoRate = 1
        if resume && isPlaying { v.play() } // may have been frozen at the end
    }

    /// Keeps `video.rate` writes to a minimum — setting it every tick is itself
    /// enough to hitch playback.
    private var videoRate: Float = 1 {
        didSet {
            guard videoRate != oldValue, let v = video, isPlaying else { return }
            v.rate = videoRate
        }
    }

    private func startDrift() {
        stopDrift()
        // Gentle and infrequent: hard-seeking a streaming video on small drift
        // costs a rebuffer stutter, so only large drift is corrected.
        driftTimer = Timer.scheduledTimer(withTimeInterval: 1, repeats: true) { [weak self] _ in
            MainActor.assumeIsolated { self?.correctDrift() }
        }
    }

    private func correctDrift() {
        guard isPlaying else { return }
        // A stalled or failed master reports a frozen time; correcting against
        // it would seek the MV back to the same spot every tick.
        guard master.currentItem?.status == .readyToPlay, master.rate > 0 else { return }
        let m = positionSec
        if canToggle {
            let other = hasInstrumental ? original : instrumental
            let delta = other.currentTime().seconds - m
            if delta.isFinite && abs(delta) > 0.4 {
                other.seek(to: CMTime(seconds: m, preferredTimescale: 600),
                           toleranceBefore: .zero, toleranceAfter: .zero)
            }
        }
        guard let v = video else { return }
        if videoEnded(at: m) { return }
        // After a jump, let the decoder settle instead of judging it mid-seek.
        guard Date() >= videoSettledAt else { return }
        let target = videoTarget(for: m)
        var delta = v.currentTime().seconds - target
        guard delta.isFinite else { return }
        if videoLoops && videoDuration > 0 {
            // Wrap-around distance: never seek across the loop seam.
            let half = videoDuration / 2
            if delta > half { delta -= videoDuration }
            if delta < -half { delta += videoDuration }
        }
        switch abs(delta) {
        case 1.2...:
            // Too far to walk back: jump, then stop judging for a moment.
            syncVideo(to: m, tolerance: 0.25, resume: false)
            videoSettledAt = Date().addingTimeInterval(2)
        case 0.15...:
            // Close enough to converge by running the picture a hair fast or
            // slow. Invisible to the eye, and it costs no decode work.
            videoRate = delta > 0 ? 0.97 : 1.03
        default:
            videoRate = 1
        }
    }

    private func stopDrift() {
        driftTimer?.invalidate()
        driftTimer = nil
    }

    // ---- end of item -------------------------------------------------------

    private func installEndObservers() {
        removeObservers()
        if let item = master.currentItem {
            observers.append(NotificationCenter.default.addObserver(
                forName: .AVPlayerItemDidPlayToEndTime, object: item, queue: .main
            ) { [weak self] _ in
                MainActor.assumeIsolated {
                    guard let self else { return }
                    self.isPlaying = false
                    self.stopDrift()
                    self.onEnded?()
                }
            })
        }
        if let v = video, let item = v.currentItem {
            observers.append(NotificationCenter.default.addObserver(
                forName: .AVPlayerItemDidPlayToEndTime, object: item, queue: .main
            ) { [weak self] _ in
                MainActor.assumeIsolated {
                    guard let self, self.videoLoops else { return }
                    v.seek(to: .zero)
                    if self.isPlaying { v.play() }
                }
            })
        }
    }

    private func removeObservers() {
        for o in observers { NotificationCenter.default.removeObserver(o) }
        observers.removeAll()
    }

    private func reset() async {
        isPlaying = false
        isLoading = false
        loadError = nil
        trackNotice = nil
        videoLoops = false
        videoDuration = 0
        videoOffsetSec = 0
        videoRate = 1
        videoSettledAt = .distantPast
        duration = 0
        stopDrift()
        fadeTimer?.invalidate()
        removeObservers()
        for p in [original, instrumental] {
            p.pause()
            p.replaceCurrentItem(with: nil)
            p.volume = 0
        }
        video?.pause()
        video?.replaceCurrentItem(with: nil)
        video = nil
        hasOriginal = false
        hasInstrumental = false
    }
}
